docs: bridging sequence diagram (#275)
agglayer/aggkit: docs/bridge_service.md

## Bridge flow

> Note that L2 networks consist of the `aggkit` node and execution client.

Reviewer: Now that we have a different participant for the `Aggkit`, we don't need this disclaimer, I'd say.
```suggestion
```

    Aggkit (B)->>L1: Aggoracle fetches last finalized GER from L1
    Aggkit (B)->>L2 (B): Aggoracle injects the GER on L2 (B) GlobalExitRootManagerL2SovereignChain.sol

    User->>Aggkit (A): Call bridge_l1InfoTreeIndexForBridge endpoint on the origin network(A)
    Aggkit (A)-->>User: Returns L1InfoTree index X for which the bridge was included
    User->>Aggkit (B): Poll bridge_injectedInfoAfterIndex on destination network until a non-null response is retreived.
    Aggkit (B)-->>User: returns a L1InfoTreeLeaf of injected GER Y from destination L2 network
    User->>Aggkit (A): Call bridge_getProof on origin network(A) to generate merkle proof for bridge using l1InfoTreeIndex of GER Y and networkID(A)
    
    Aggkit (A)-->>User: Return claim proof
    User->>L2 (B): Claim (proof)
    L2 (B)->>L2 (B): Send claim tx<br/>(bridge is settled on the L2 (B))
    L2 (B)-->>User: Tx hash
```

### Bridge flow L1 -> L2

The diagram below describes the basic L1 -> L2 bridge workflow.

```mermaid
sequenceDiagram
    participant User
    participant L1
    participant Aggkit
    participant L2

    User->>L1: Bridge assets to L2
    L1->>L1: Updates the mainnet exit tree

>     L1->>L1: update GER

Reviewer: ```suggestion
    L1->>L1: Update GER
```

    Note right of L1: bridgeContract updates the GER<br/>only if `forceUpdateGlobalExitRoot` is true in the bridge transaction.
    Aggkit->>L1: Aggoracle fetches last finalized GER
    Aggkit->>L2: Aggoracle injects the GER on L2 GlobalExitRootManagerL2SovereignChain.sol

    User->>Aggkit: Call bridge_l1InfoTreeIndexForBridge endpoint on the origin network
    Aggkit-->>User: Returns L1InfoTree index X for which the bridge was included
    User->>Aggkit: Poll bridge_injectedInfoAfterIndex on destination network until a non-null response is retreived.

>     Aggkit-->>User: returns a L1InfoTreeLeaf of injected GER Y from destination L2 network

Reviewer: Use the loop notation here as well.

    User->>Aggkit: Call bridge_getProof on origin network to generate merkle proof for bridge using l1InfoTreeIndex of GER Y and networkID=0 (L1)
    Aggkit-->>User: Return claim proof
    User->>L2: Claim (proof)
    L2->>L2: Send claimAsset/claimBridge tx on the destination network<br/>(bridge is settled on the L2)
    L2-->>User: Tx hash
```

**Notes:**  

> 1. In CDK-Erigon, the Global Exit Root (GER) on the L2 smart contract (`PolygonZKEVMGlobalExitRootL2.sol`) is automatically updated by the sequencer. In a sovereign chain, the GER is injected on L2 by the Aggoracle component.  

Reviewer: ```suggestion
1. In CDK-Erigon, the Global Exit Root (GER) on the L2 smart contract (`PolygonZkEVMGlobalExitRootV2.sol`) is automatically updated by the sequencer. In a sovereign chain, the GER is injected on L2 (`GlobalExitRootManagerL2SovereignChain.sol`) by the Aggoracle component.  
```

Author: should not this be `PolygonZKEVMGlobalExitRootL2` instead of `PolygonZkEVMGlobalExitRootV2` ?
As `PolygonZkEVMGlobalExitRootV2` is deployed on L1 for managing GER

Author: AFAIK current stack also utilizes the contracts outside of `v2` folder. `PolygonZKEVMGlobalExitRootL2` is used for managing and storing GER on L2 and `PolygonZkEVMGlobalExitRootV2` is used for managing GER on L1 amongst other things.

Correct me if i am mistaken here

Author: As discussed resolving this for now

### Bridge flow L2 -> L1

The diagram below describes the basic L2 -> L1 bridge workflow.

```mermaid
sequenceDiagram
    participant User
    participant L2
    participant Aggkit
    participant AggLayer
    participant L1

    User->>L2: Bridge assets to L1
    L2->>L2: Index bridge tx & updates the local exit tree
    Aggkit->>AggLayer: Build & send certificate (Aggsender)
    AggLayer->>L1: Settle batch
    L1->>L1: update GER
    Note right of L1: rollupmanager updates the GER & RER (PolygonZKEVMGlobalExitRootV2.sol)
    AggLayer-->>L2: Return L1 tx hash
    Aggkit->>L1: Fetch last finalized GER (Aggoracle)
    Aggkit->>L2: Aggoracle injects GER on L2 (GlobalExitRootManagerL2SovereignChain.sol)

    User->>Aggkit: Query bridge_l1InfoTreeIndexForBridge endpoint on the origin network(L2)
    Aggkit-->>User: Returns L1InfoTree index X for which the bridge was included 
    User->>Aggkit: Poll bridge_injectedInfoAfterIndex on destination network(L1) until a non-null response.

>     User->>Aggkit: Returns the first L1InfoTreeLeaf(GER=Y) for the GER injected at or after L1InfoTree index X

Reviewer: Use the loop notion here, wdyt?
```suggestion
    loop Poll destination network, until `L1InfoTreeLeaf` is retrieved
      User->>Aggkit: Poll bridge_injectedInfoAfterIndex on destination network (L1) until a non-null response.
      Aggkit-->>User: Returns the first L1InfoTreeLeaf(GER=Y) for the GER injected at or after L1InfoTree index X
    end
```

L2 (B)-->>User: Tx hash
```

### Bridge flow L1 -> L2

The diagram below describes the basic L1 -> L2 bridge workflow.

```mermaid
sequenceDiagram
    participant User
    participant L1
    participant Aggkit
    participant L2

    User->>L1: Bridge assets to L2
    L1->>L1: Updates the mainnet exit root
    L1->>L1: bridgeContract updates the GER in PolygonZKEVMGlobalExitRootV2.sol<br/>if forceUpdateGlobalExitRoot is true in the bridge transaction
    Aggkit->>L1: Aggoracle fetches last finalized GER
    Aggkit->>L2: Aggoracle injects the GER on L2 GlobalExitRootManagerL2SovereignChain.sol

    User->>Aggkit: Call bridge_l1InfoTreeIndexForBridge endpoint on the origin network
    Aggkit-->>User: Returns L1InfoTree index X for which the bridge was included
    User->>Aggkit: Poll bridge_injectedInfoAfterIndex on destination network until a non-null response is retreived.

>     User-->>Aggkit: Get first GER injected that happened at or after L1InfoTree index X

Reviewer: ```suggestion
    User->>Aggkit: Get first GER injected that happened at or after L1InfoTree index X
```

>     L2-->>User: GER Y

Reviewer: ```suggestion
    Aggkit-->>User: return injected GER Y from destination L2 network
```

    User->>Aggkit: Call bridge_getProof on origin network to generate merkle proof for bridge using GER Y and networkID=0 (L1)

>     Aggkit-->>User: Proof

Reviewer: ```suggestion
    Aggkit-->>User: Return claim proof
```
